cirq-core/cirq/transformers/routing/route_circuit_cqc.py

```diff
@@ -0,0 +1,327 @@
+
+    def _route(
+        self,
+        timesteps: List[List['cirq.Operation']],
+        single_qubit_ops: List[List['cirq.Operation']],
+        max_search_radius: int,
+        tag_inserted_swaps: bool = False,
+    ) -> List[List['cirq.Operation']]:
+        """Main routing procedure that creates the routed circuit, inserts the user's gates in it,
+        and inserts the necessary swaps.
+
+        Args:
+          timesteps: the circuit's timesteps as defined by the paper
+          max_search_radius: the maximum number of times the cost function can be iterated for
+            convergence.
+
+        Returns:
+          a list of lists corresponding to moments of the routed circuit
+        """
+
+        def process_executable_ops(idx: int):
+            unexecutable_ops = []
+            for op in timesteps[idx]:
+                if self._mm.can_execute(op):
+                    routed_ops[idx].append(self._mm.mapped_op(op))
+                else:
+                    unexecutable_ops.append(op)
+            timesteps[idx] = unexecutable_ops
+
+        routed_ops: List[List['cirq.Operation']] = [[] for i in range(len(timesteps) + 1)]
+        for idx in range(len(timesteps)):
+            # add single qubit ops the current output moment.
+            routed_ops[idx].extend([self._mm.mapped_op(op) for op in single_qubit_ops[idx]])
+
+            process_executable_ops(idx)
+
+            seen: Set[Tuple[Tuple['cirq.Qid', cirq.Qid], ...]] = set()
+            while len(timesteps[idx]) != 0:
+                sigma = self._initial_candidate_swaps(timesteps[idx])
+                for s in range(idx, min(max_search_radius + idx, len(timesteps))):
+                    if len(sigma) <= 1:
+                        break
+                    sigma = self._next_candidate_swaps(sigma, timesteps[s])
+
+                if len(sigma) > 1 and idx + max_search_radius <= len(timesteps):
+                    chosen_swaps = self._symmetry_swap_pair(timesteps, idx, max_search_radius)
+                else:
+                    chosen_swaps = sigma[0]
+
+                if chosen_swaps in seen:
+                    chosen_swaps = self._symmetry_brute_force(timesteps, idx)
+                else:
+                    seen.add(chosen_swaps)
+
+                for swap in chosen_swaps:
+                    inserted_swap = self._mm.mapped_op(ops.SWAP(*swap))
+                    if tag_inserted_swaps:
+                        inserted_swap.with_tags(ops.RoutingSwapTag())
+                    routed_ops[idx].append(inserted_swap)
+                    self._mm.apply_swap(*swap)
+                process_executable_ops(idx)
+
+        # edge case: single qubit gates may appear in moments after all 2-qubit gates
+        for i in range(len(timesteps), len(single_qubit_ops)):
+            routed_ops.append([self._mm.mapped_op(op) for op in single_qubit_ops[i]])
+
+        return routed_ops
+
+    def _disjoint_qubit_combinations(
```

Comment:
consider making this a free function (or a `@staticmethod`, but usually those should be free functions)

Reply:
Sounds good. I'll make it a free function.